### PR TITLE
[ROCm][Perf] Pad fp16/bf16 GEMM weight row stride for L2 cache

### PR DESCRIPTION
**Pull request by mgehre-amd**
On gfx11 (RDNA3/3.5) a dense GEMM whose weight row stride is a multiple of 2048 bytes collide on the L2/MALL hashes, costing a large share of achievable bandwidth. Offsetting the row stride by one 128 B cache line dodges the collision.
This PR improves TTFT by up to 13% and TPOT by up to 3%.

We pad the row stride once in `UnquantizedLinearMethod.process_weights_after_loading`.
It has a minimal impact on memory consumption on affected layers only (1.29 % of weight memory for Qwen3-8B: 171 MiB on 12.94 GiB).

The predicate is `(K * element_size) % 2048 == 0`, so for bf16/fp16 it fires on any layer
with `K % 1024 == 0`. That covers most of the Llama/Qwen dense family: for Qwen3-8B all
four linears qualify (`qkv`/`o`/`gate_up` `K=4096` → 8192 B, `down` `K=12288` → 24576 B).

Gated on gfx11. This might also apply to gfx12, but we will need to measure and can enable as a follow up.

## Benchmark

On AMD Strix Halo (gfx1151),
`--num-prompts 5`, 2 interleaved reps per cell. Both arms only differ in `VLLM_ROCM_LINEAR_PADDING`.

| Model (in/out len) | Qualifying | Median TTFT | Median TPOT | Weight memory |
|---|---|---|---|---|
| `Qwen3-8B` 1920/128 | 100 % | 918.02 → 852.39 ms (**−7.15 %**) | 72.15 → 69.83 ms (**−3.21 %**) | 15.27 → 15.47 GiB |
| `Qwen3-8B` 128/128 | 100 % | 119.35 → 103.59 ms (**−13.20 %**) | 70.50 → 68.11 ms (**−3.39 %**) | 15.27 → 15.47 GiB |
| `Qwen2.5-VL-3B-Instruct` 1920/128 | 70.7 % | 969.01 → 932.10 ms (**−3.81 %**) | 30.37 → 29.36 ms (**−3.33 %**) | 7.16 → 7.27 GiB |
| `Qwen3-4B` 1920/128 | 10.4 % | 512.12 → 510.49 ms (not resolvable) | 38.85 → 38.03 ms (**−2.11 %**) | 7.56 → 7.57 GiB |
| `Qwen2-1.5B-Instruct` 1920/128 | 0 % | 198.34 → 201.03 ms (no effect) | 14.64 → 14.74 ms (no effect) | 2.89 → 2.89 GiB |

## Accuracy

`tests/evals/gsm8k/gsm8k_eval.py`, full 1319-question test set, 5-shot, greedy
(temperature 0.0, seed 42), `Qwen/Qwen3-8B` on gfx1151:

| | Accuracy | Correct | Invalid |
|---|---|---|---|
| padding off | 88.931 % | 1173 / 1319 | 0.0 |
| padding on | 89.159 % | 1176 / 1319 | 0.0 |

## AI assistance

AI assistance was used for this PR. Every changed line was reviewed by the submitter, and
all tests, benchmarks and evals listed above were run on the hardware named.

### CODE REVIEW

**Review by tjtanaa — CHANGES_REQUESTED** [review]
@mgehre-amd please keep all your changes confined to `gfx11/Strix Halo`. Right now it is affecting all GPU archs, even the instinct.

**Comment by mgehre-amd** [comment]
> @mgehre-amd please keep all your changes confined to `gfx11/Strix Halo`. Right now it is affecting all GPU archs, even the instinct.

@tjtanaa, `cache_aliasing_pad_elems` uses 
```CACHE_ALIASING_GEOMETRY: dict[str, tuple[int, int]] = {
    "gfx11": (2048, 128),
}
```
(https://github.com/ROCm/vllm/blob/2dae1c74b97702aa81a9be17185b90b5b0079bbe/vllm/model_executor/layers/utils.py#L265-L267)

so it only applies padding on gfx11 and keeps all others archs unchanged.

**Comment by wjabbour** [comment]
@mgehre-amd  Agreeing with TJ here. You're right that `cache_aliasing_pad_elems` itself only pads on gfx11, but the operand-compatibility checks you changed in rocm_unquantized_gemm_impl (utils.py)  aren't gated the same way and run on every ROCm arch:
 - `wvsplitk_operands_compatible = weight.stride(1) == 1 and bias_compatible` relaxes the old `weight.is_contiguous()` gate for wvSplitK everywhere, not just where padding can occur
 - the new `and weight.is_contiguous()` added to the `aiter-triton-gemm` and `tgemm` branches is a new restriction on every arch, including instinct
 
Neither is inside your padding helper or checked against gfx11, so instinct's kernel selection can change even though its weights are never padded. Can you gate those behind the same gfx11 check (or only apply the relaxed check when the weight was actually padded)?

**Comment by mgehre-amd** [comment]
> @mgehre-amd Agreeing with TJ here. You're right that `cache_aliasing_pad_elems` itself only pads on gfx11, but the operand-compatibility checks you changed in rocm_unquantized_gemm_impl (utils.py) aren't gated the same way and run on every ROCm arch:

Sure, makes sense! I updated this PR to keep gfx9 exactly the same as before. The changes to the guards are minimized and the important one reads
````
# wvSplitK allows row strides. Perf impact is only validated on gfx11.
and (weight.is_contiguous() or (on_gfx11() and weight.stride(1) == 1))
```
so we will only allow non-contiguous strides on gfx11.

File: tests/model_executor/layers/test_rocm_unquantized_gemm.py [hunk 1]
```diff
@@ -149,6 +256,24 @@ def test_rocm_unquantized_gemm_noncontiguous_activation_real_kernel(monkeypatch,
     torch.testing.assert_close(out, ref, atol=1e-2, rtol=1e-2)
 
 
+@pytest.mark.skipif(not current_platform.is_rocm(), reason="ROCm-only kernel test")
+@pytest.mark.parametrize("dtype", [torch.float16, torch.bfloat16])
+def test_rocm_unquantized_gemm_padded_weight_real_kernel(monkeypatch, dtype):
```
Comment (wjabbour) [inline tests/model_executor/layers/test_rocm_unquantized_gemm.py:272]: this test should only run on gfx11, else it'll get picked up by a CDNA runner and fail

File: tests/model_executor/layers/test_rocm_unquantized_gemm.py [hunk 2]
```diff
@@ -130,6 +130,113 @@ def test_rocm_unquantized_gemm_rejects_unsupported_skinny_layouts(
...
+def _row_stride_padded(weight: torch.Tensor, pad: int) -> torch.Tensor:
+    """A weight whose row stride is offset by ``pad`` elements, as produced by
+    ``maybe_pad_weight_avoid_cache_aliasing``."""
+    padded = torch.nn.functional.pad(weight, (0, pad))[..., :-pad]
+    assert padded.shape == weight.shape
+    assert padded.stride() == (weight.shape[1] + pad, 1)
+    return padded
+
+
+def test_rocm_unquantized_gemm_wvsplitk_accepts_row_stride_padded_weight(monkeypatch):
```
Comment (wjabbour) [inline tests/model_executor/layers/test_rocm_unquantized_gemm.py:151]: this is missing `monkeypatch.setattr("vllm.platforms.rocm.on_gfx11", lambda: True)` so that it can execute successfully on the CDNA runners

File: vllm/model_executor/layers/utils.py [hunk 3]
```diff
@@ -323,7 +373,9 @@ def rocm_unquantized_gemm_impl(
         # TODO GFX1250: Include once skinny GEMM is supported on gfx1250
         and x.dtype in [torch.float16, torch.bfloat16]
         and k % 8 == 0
-        and skinny_operands_compatible
+        and (bias is None or bias.is_contiguous())
```
Comment (wjabbour) [inline vllm/model_executor/layers/utils.py:403]: nit:

The routing behavior here is correctly scoped to gfx11 as far as I can trace it, but the code is creating this boolean condition, `use_skinny`, and then re-evaluating the conditions inside the block, which is confusing to reason about. 

This might be a bit chunky for this PR, but I recommend breaking the conditions down into separate checks and composing them, for clarity

e.g.
```
skinny_gemm_enabled = (
    envs.VLLM_ROCM_USE_SKINNY_GEMM
    and (on_gfx9() or on_gfx1x())
    and x.dtype in (torch.float16, torch.bfloat16)
    and k % 8 == 0
)

bias_ok = bias is None or bias.is_contiguous()
weight_dense = weight.is_contiguous()

# wvSplitK reads the weight row stride, so a gfx11 cache-aliasing-padded
# (K-contiguous) weight is fine; LLMM1 indexes off a raw pointer and needs dense.
weight_ok_for_wvsplitk = weight_dense or (on_gfx11() and weight.stride(1) == 1)

use_wvsplitk = (
    skinny_gemm_enabled and weight_ok_for_wvsplitk and bias_ok
    and m > 8 and 0 < n <= 5
)

use_llmm1 = (
    skinny_gemm_enabled and weight_dense and bias is None
    and m % 4 == 0 and n == 1 and k <= 8192
)

if use_wvsplitk or use_llmm1:
    x_view = x.reshape(-1, x.size(-1)).contiguous()
    if use_wvsplitk:
        out = ops.wvSplitK(weight, x_view, num_compute_units(), bias)
        return out.reshape(*x.shape[:-1], weight.shape[0])
    out = ops.LLMM1(weight, x_view, 4)                                         
    return out.reshape(*x.shape[:-1], weight.shape[0])
```

File: tests/model_executor/layers/test_rocm_unquantized_gemm.py (hunk 1 above)
Comment (mgehre-amd) [reply tests/model_executor/layers/test_rocm_unquantized_gemm.py:272]: thanks, fixed!

File: tests/model_executor/layers/test_rocm_unquantized_gemm.py (hunk 2 above)
Comment (mgehre-amd) [reply tests/model_executor/layers/test_rocm_unquantized_gemm.py:151]: thanks, fixed!

File: vllm/model_executor/layers/utils.py (hunk 3 above)
Comment (mgehre-amd) [reply vllm/model_executor/layers/utils.py:403]: Thanks, I simplified the logic following your suggestions!

**Review by wjabbour — APPROVED** [review]
appreciate your work here, Matthias!

File: vllm/model_executor/layers/utils.py [hunk 4]
```diff
@@ -333,27 +384,46 @@ def rocm_unquantized_gemm_impl(
...
-        if (m == 1 or m > 8) and 0 < n <= 5:
-            cu_count = num_compute_units()
+        if use_wvsplitk:
             out = ops.wvSplitK(weight, x_view, cu_count, bias)
-            return out.reshape(*x.shape[:-1], weight.shape[0])
-        elif m % 4 == 0 and n == 1 and k <= 8192 and bias is None:
+        else:
             out = ops.LLMM1(weight, x_view, 4)
-            return out.reshape(*x.shape[:-1], weight.shape[0])
+        return out.reshape(*x.shape[:-1], weight.shape[0])
```
Comment (roberteg16) [inline vllm/model_executor/layers/utils.py]: Perhaps I am mistaken here, but if use_llmm1 is false, `out = ops.LLMM1(weight, x_view, 4)` will be called anyway. 

Is that intended?
